yt_dlp/extractor/kenh14.py

```diff
@@ -0,0 +1,150 @@
+)
+from ..utils.traversal import traverse_obj
+
+
+class Kenh14VideoIE(InfoExtractor):
+    _VALID_URL = r'https?://video\.kenh14\.vn/(?:video/)?[\w-]+-(?P<id>[0-9]+)\.chn'
+    _TESTS = [{
+        'url': 'https://video.kenh14.vn/video/mo-hop-iphone-14-pro-max-nguon-unbox-therapy-316173.chn',
+        'md5': '1ed67f9c3a1e74acf15db69590cf6210',
+        'info_dict': {
+            'id': '316173',
+            'ext': 'mp4',
+            'title': 'Video mở hộp iPhone 14 Pro Max (Nguồn: Unbox Therapy)',
+            'description': 'Video mở hộp iPhone 14 Pro MaxVideo mở hộp iPhone 14 Pro Max (Nguồn: Unbox Therapy)',
+            'thumbnail': r're:^https?://videothumbs\.mediacdn\.vn/.*\.jpg$',
+            'tags': [],
+            'uploader': 'Unbox Therapy',
+            'upload_date': '20220517',
+            'view_count': int,
+            'duration': 722.86,
+            'timestamp': 1652764468,
+        },
+    }, {
+        'url': 'https://video.kenh14.vn/video-316174.chn',
+        'md5': '2b41877d2afaf4a3f487ceda8e5c7cbd',
+        'info_dict': {
+            'id': '316174',
+            'ext': 'mp4',
+            'title': 'Khoảnh khắc VĐV nằm gục khóc sau chiến thắng: 7 năm trời Việt Nam mới có HCV kiếm chém nữ, chỉ có 8 tháng để khổ luyện trước khi lên sàn đấu',
+            'description': 'md5:de86aa22e143e2b277bce8ec9c6f17dc',
+            'thumbnail': r're:^https?://videothumbs\.mediacdn\.vn/.*\.jpg$',
+            'tags': [],
+            'upload_date': '20220517',
+            'view_count': int,
+            'duration': 70.04,
+            'timestamp': 1652766021,
+        },
+    }, {
+        'url': 'https://video.kenh14.vn/0-344740.chn',
+        'md5': 'b843495d5e728142c8870c09b46df2a9',
+        'info_dict': {
+            'id': '344740',
+            'ext': 'mov',
+            'title': 'Kỳ Duyên đầy căng thẳng trong buổi ra quân đi Miss Universe, nghi thức tuyên thuệ lần đầu xuất hiện gây nhiều tranh cãi',
+            'description': 'md5:2a2dbb4a7397169fb21ee68f09160497',
+            'thumbnail': r're:^https?://kenh14cdn\.com/.*\.jpg$',
+            'tags': ['kỳ duyên', 'Kỳ Duyên tuyên thuệ', 'miss universe'],
+            'uploader': 'Quang Vũ',
+            'upload_date': '20241024',
+            'view_count': int,
+            'duration': 198.88,
+            'timestamp': 1729741590,
+        },
+        'expected_warnings': [
+            'Failed to download m3u8 information: HTTP Error 404: NOT FOUND',
+        ],
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        attrs = extract_attributes(get_element_html_by_attribute('type', 'VideoStream', webpage) or '')
+        direct_url = attrs['data-vid']
+
+        metadata = self._download_json(
+            'https://api.kinghub.vn/video/api/v1/detailVideoByGet?FileName={}'.format(
+                remove_start(direct_url, 'kenh14cdn.com/')), video_id, fatal=False)
+
```

Comment:
imo. we should use just the new api instead of constructing the m3u8
```suggestion

        formats = [{'url': f'https://{direct_url}', 'format_id': 'http'}]
        subtitles = {}
        video_data = self._download_json(f'https://{direct_url}.json', video_id, fatal=False)
        if video_data:
            if hls_url := video_data.get('hls'):
                fmts, subs = self._extract_m3u8_formats_and_subtitles(
                    hls_url, video_id, m3u8_id='hls', fatal=False)
                formats.extend(fmts)
                self._merge_subtitles(subs, target=subtitles)
            if dash_url := video_data.get('mpd'):
                fmts, subs = self._extract_mpd_formats_and_subtitles(
                    dash_url, video_id, mpd_id='dash', fatal=False)
                formats.extend(fmts)
                self._merge_subtitles(subs, target=subtitles)

```

Reply:
The patch:

```patch
From bff8c77d1cb1394822407121dafd5f77b71bcf46 Mon Sep 17 00:00:00 2001
From: Mozi <29089388+pzhlkj6612@users.noreply.github.com>
Date: Sun, 17 Nov 2024 04:01:04 +0000
Subject: [PATCH] use new api instead of constructing m3u8; extract subtitles

Co-Authored-By: sepro <sepro@sepr0.com>
---
 yt_dlp/extractor/kenh14.py | 25 +++++++++++++++++++------
 1 file changed, 19 insertions(+), 6 deletions(-)

diff --git a/yt_dlp/extractor/kenh14.py b/yt_dlp/extractor/kenh14.py
index be66adb1f..5af736930 100644
--- a/yt_dlp/extractor/kenh14.py
+++ b/yt_dlp/extractor/kenh14.py
@@ -52,10 +52,10 @@ class Kenh14VideoIE(InfoExtractor):
         },
     }, {
         'url': 'https://video.kenh14.vn/0-344740.chn',
-        'md5': 'b843495d5e728142c8870c09b46df2a9',
+        'md5': '3e3704177c3e8af5ad389d8d66d332d9',
         'info_dict': {
             'id': '344740',
-            'ext': 'mov',
+            'ext': 'mp4',
             'title': 'Kỳ Duyên đầy căng thẳng trong buổi ra quân đi Miss Universe, nghi thức tuyên thuệ lần đầu xuất hiện gây nhiều tranh cãi',
             'description': 'md5:2a2dbb4a7397169fb21ee68f09160497',
             'thumbnail': r're:^https?://kenh14cdn\.com/.*\.jpg$',
@@ -78,6 +78,21 @@ def _real_extract(self, url):
         attrs = extract_attributes(get_element_html_by_attribute('type', 'VideoStream', webpage) or '')
         direct_url = attrs['data-vid']
 
+        formats = [{'url': f'https://{direct_url}', 'format_id': 'http'}]
+        subtitles = {}
+        video_data = self._download_json(f'https://{direct_url}.json', video_id, fatal=False)
+        if video_data:
+            if hls_url := video_data.get('hls'):
+                fmts, subs = self._extract_m3u8_formats_and_subtitles(
+                    hls_url, video_id, m3u8_id='hls', fatal=False)
+                formats.extend(fmts)
+                self._merge_subtitles(subs, target=subtitles)
+            if dash_url := video_data.get('mpd'):
+                fmts, subs = self._extract_mpd_formats_and_subtitles(
+                    dash_url, video_id, mpd_id='dash', fatal=False)
+                formats.extend(fmts)
+                self._merge_subtitles(subs, target=subtitles)
+
         metadata = self._download_json(
             'https://api.kinghub.vn/video/api/v1/detailVideoByGet?FileName={}'.format(
                 remove_start(direct_url, 'kenh14cdn.com/')), video_id, fatal=False)
@@ -94,10 +109,8 @@ def _real_extract(self, url):
                 traverse_obj(metadata, ('title', {strip_or_none}))
                 or clean_html(self._og_search_title(webpage))
                 or clean_html(get_element_by_class('vdbw-title', webpage))),
-            'formats': [
-                {'url': f'https://{direct_url}', 'format_id': 'http'},
-                *self._extract_m3u8_formats(f'https://{direct_url}/master.m3u8', video_id, fatal=False),
-            ],
+            'formats': formats,
+            'subtitles': subtitles,
             'description': (
                 clean_html(self._og_search_description(webpage))
                 or clean_html(get_element_by_class('vdbw-sapo', webpage))),
-- 
2.44.0.windows.1


```